src/matrixrtc/NewMembershipManager.ts

```diff
@@ -0,0 +1,918 @@
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Check if we have a rate limit error and schedule the same action again if we dont exceed the rate limit retry count yet.
+     * @param error the error causing this handler check/execution
+     * @param method the method used for the throw message
+     * @param type which MembershipActionType we reschedule because of a rate limit.
+     * @throws If it is a rate limit error and the retry count got exceeded
+     * @returns Returns true if we handled the error by rescheduling the correct next action.
+     * Returns false if it is not a network error.
+     */
+    private rateLimitErrorHandler(error: unknown, method: string, type: MembershipActionType): boolean {
+        // "Is rate limit"-boundary
+        if (!(error instanceof HTTPError && error.isRateLimitError())) {
+            return false;
+        }
+
+        // retry boundary
+        const rateLimitRetries = this.scheduler.state.rateLimitRetries.get(type) ?? 0;
+        if (rateLimitRetries < this.maximumRateLimitRetryCount) {
+            let resendDelay: number;
+            const defaultMs = 5000;
+            try {
+                resendDelay = error.getRetryAfterMs() ?? defaultMs;
+                logger.info(`Rate limited by server, retrying in ${resendDelay}ms`);
+            } catch (e) {
+                logger.warn(
+                    `Error while retrieving a rate-limit retry delay, retrying after default delay of ${defaultMs}`,
+                    e,
+                );
+                resendDelay = defaultMs;
+            }
+            this.scheduler.state.rateLimitRetries.set(type, rateLimitRetries + 1);
+            this.scheduler.addAction({ ts: Date.now() + resendDelay, type });
+            return true;
+        }
+
+        throw Error("Exceeded maximum retries for " + type + " attempts (client." + method + "): " + (error as Error));
+    }
+
+    /**
+     * FIXME Don't Check the error and retry the same MembershipAction again in the configured time and for the configured retry count.
+     * @param error the error causing this handler check/execution
+     * @param type the action type that we need to repeat because of the error
+     * @throws If it is a network error and the retry count got exceeded
+     * @returns
+     * Returns true if we handled the error by rescheduling the correct next action.
+     * Returns false if it is not a network error.
+     */
+    private retryOnNetworkError(error: unknown, type: MembershipActionType): boolean {
+        // "Is a network error"-boundary
+        const retries = this.scheduler.state.networkErrorRetries.get(type) ?? 0;
+        const retryDurationString = this.callMemberEventRetryDelayMinimum / 1000 + "s";
+        const retryCounterString = "(" + retries + "/" + this.maximumNetworkErrorRetryCount + ")";
+        if (error instanceof Error && error.name === "AbortError") {
```

Comment:
As per DM, I don't think this will detect network errors when EC is running as embedded widget.

I *think* we would see `MatrixError`s instead which we would need to parse.